### ISSUE TITLE
Python: Bug: Issues with Anthropic tool calling

### ISSUE THREAD

**Issue by nikdavis** [microsoft/semantic-kernel#9853]
**Describe the bug**
I should probably split these out, lmk if that would be helpful.

There are a couple issues with Anthropic tool calling. AFAICT both issues are in 
`python/semantic_kernel/connectors/ai/anthropic/services/anthropic_chat_completion.py`

1)
if an agent self-invokes tools more than 5 times (`DEFAULT_MAX_AUTO_INVOKE_ATTEMPTS`) the tool list will be removed from the last chat message (assuming to force stop auto invokes?)
This may work for some APIs but it causes Anthropic's API to error out:
`Requests which include tool_use or tool_result blocks must define tools.`
If it sees tool history, it expects tool calls .. at least that's my interpretation.

2)
`_get_tool_calls_from_message` does not always yield hashable `FunctionCallContent`s. It passes a `dict` for the `arguments` param, which is not hashable. The FCCs seem to be serialized between group chat handoffs (?) which forces the error. My workaround has been to serialize the tool call arguments in `_get_tool_calls_from_message` which seems to be working fine. Can PR if that's helpful, but not totally sure what the expected behavior is here.

**To Reproduce**

1)
- init an anthropic service
- init an agent with that service, setting `execution_settings.function_choice_behavior = FunctionChoiceBehavior.Auto()`
- ask it to do something complex enough that it does >=5 tool calls in a row (or override to 2?)

2)
This one is a little tricker to have happen and haven't pinned down the exact case to repro, but it's something along the lines of
- initialize an AgentGroupChat using an anthropic service
- ask to do something that requires tool use
- on handoff (either to user or to another assistant?) tool calls are serialized (maybe for ChatHistory? not sure) and I think that is likely what forces the issue

In general it seems clear though that `FunctionCallContent`s are intended to be hashable and the `AnthropicChatCompletion` does not meet that goal. Not clear to me why `FunctionCallContent` allows a `dict` for the `arguments` param (maybe legacy?) .. e.g. type -> `arguments: str | dict[str, Any] | None = None`

**Expected behavior**

1)
- not totally sure. I think the intended behavior (assuming) is to backoff the agent from self-invoking more than 5 tool calls. Maybe there is a more anthropic-ey way of handling this than just blowing away the tool call part of the payload (assuming I've debugged this correctly)

2)
- somehow the `arguments` parameter needs to be hashable – maybe it's just okay to `json.dumps` ?

**Screenshots**

**Platform**
 - OS: Mac Seqouia 15.1
 - Language: Python
 - Source: 1.16.0 / using uv

**Additional context**

**Comment by evchaki** [comment]
@nikdavis - thanks for logging this.  @eavanvalkenburg can you take a look at this?

**Comment by TaoChenOSU** [comment]
Currently looking for a way to stop Anthropic from invoking functions when the tool definitions cannot be taken out from the request.

**Comment by TaoChenOSU** [comment]
I am able to reproduce the issue. Anthropic expects tools in the request when the chat history contains tool use. This is incompatible with how we are handling max auto invocation attempts in Python. However, in .Net, when the max auto invocation attempts have been reached, the model can still pick a tool in the response, essentially falling back to manually invocation mode. This is a discrepancy between .Net and Python and we need to address this.

**Comment by TaoChenOSU** [comment]
The second issue should have been resolved by this PR: https://github.com/microsoft/semantic-kernel/pull/9938.
The issue was happening because the Anthropic connector was not serializing the Anthropic responses into SK contents. It was relying on the `inner_content` property, which wouldn't work if the contents are given to another connector.

The first issue is a bit trickier. There is a misalignment on how we handle auto invocation between SK .Net and SK Python, which we need to address. In .Net, after the max attempt has been reached, it simply falls back to manual invocation whereas in Python, it forces the model to generate a text response.

I have tried the following to get around the misalignment to unblock people: replace the original list of tools with a fake tool to try to make the Anthropic service stop complaining. However, the model seems to be able to guess a function from the chat history. For example, it used a `math-add` function before and now it's able to guess a `math-subtract` function even there is not such function in the new list of tools.

For now, the temporary workaround is to increase the limit on max attempt. This will prevent the auto invocation loop from entering the last attempt.

**Comment by moonbox3** [comment]
Closing as #9938 was merged to main a while ago.
